src/serverlist.cpp

```diff
@@ -649,6 +649,14 @@ void CServerListManager::RetrieveAll ( const CHostAddress& InetAddr )
         // if the client IP address is a private one, it's on the same LAN as the directory
         bool clientIsInternal = NetworkUtil::IsPrivateNetworkIP ( InetAddr.InetAddr );
 
+        CHostAddress clientPublicAddr = CHostAddress ( InetAddr.InetAddr, InetAddr.iPort );
+        if ( clientIsInternal && CHostAddress().InetAddr != ServerList[0].LHostAddr.InetAddr &&
+             !NetworkUtil::IsPrivateNetworkIP ( ServerList[0].LHostAddr.InetAddr ) )
+        {
+            // client and directory on same LAN, directory has public IP set, use that for client public IP
```

Comment:
For some mad crazy reason, in this version of serverlist.cpp, LHostAddr is not the local host address, it's the Public IP for the directory.  This is because the directory doesn't swap it's HostAddr and LHostAddr when creating its server list entry...

Reply:
`LHostAddr` should probably be renamed to something more generic such as `AlternateHostAddr`?

Reply:
Not really - it holds the local host address when sending the protocol message.  (Also, it's easy enough to fix the directory to store its own details the "right" way around.  And, indeed, I have done that... I could submit a patch... I've rather a lot of code changes stacked up in a single commit currently...  I should break them up...)